cedar-policy-core/src/evaluator/err.rs

```diff
@@ -115,25 +116,46 @@ impl EvaluationError {
     }
 
     /// Construct a [`TypeError`] error
-    pub(crate) fn type_error(expected: Vec<Type>, actual: Type) -> Self {
+    pub(crate) fn type_error(expected_first: Type, expected_rest: Vec<Type>, actual: Type) -> Self {
```

Comment:
it's worth a comment on this function specifying what `expected_first` and `expected_rest` are.

I think we should consider using https://crates.io/crates/nonempty instead of essentially rolling it ourselves, if only to make code like this (and below) clearer and more ergonomic.